#### Pandas-DataFrame And Series
Pandas is a powerful data manipulation library in Python, widely used for data analysis and data cleaning. It provides two primary data structures: Series and DataFrame. A Series is a one-dimensional array-like object, while a DataFrame is a two-dimensional, size-mutable, and potentially heterogeneous tabular data structure with labeled axes (rows and columns).

In [2]:
import pandas as pd

In [2]:
## Series
##A Pandas Series is a one-dimensional array-like object that can hold any data type. It is similar to a column in a table.

import pandas as pd
data=[1,2,3,4,5,"hello"]#Data type is an object now as it is a mix of integers and string
# It does not covert all the data to string implicitely as was the case with numpy array
series=pd.Series(data)
print("Series \n",series)
print(type(series))

Series 
 0        1
1        2
2        3
3        4
4        5
5    hello
dtype: object
<class 'pandas.core.series.Series'>


In [5]:
## Create a Series from dictionary
data={'a':1,'b':2,'c':3}
series_dict=pd.Series(data)
print(series_dict)


a    1
b    2
c    3
dtype: int64


In [6]:
data=[10,20,30]
index=['a','b','c']
pd.Series(data,index=index)

a    10
b    20
c    30
dtype: int64

In [3]:
data=[10,20,30]
index=['a','b','c','d']
pd.Series(data,index=index)

ValueError: Length of values (3) does not match length of index (4)

In [4]:
data=[10,20,30,40]
index=['a','b','c']
pd.Series(data,index=index)

ValueError: Length of values (4) does not match length of index (3)

### In the above way the no of elements is data and index list need to be same ,if any of the two have different no of elements ,then we get an error

In [5]:
## Dataframe
## create a Dataframe from a dictionary of list

data={
    'Name':['Krish','John','Jack'],
    'Age':[25,30,45],
    'City':['Bangalore','New York','Florida']
}
df=pd.DataFrame(data)
print(df)
print(type(df))

    Name  Age       City
0  Krish   25  Bangalore
1   John   30   New York
2   Jack   45    Florida
<class 'pandas.core.frame.DataFrame'>


In [8]:
import numpy as np
data =np.array(df)
print(data)

[['Krish' 25 'Bangalore']
 ['John' 30 'New York']
 ['Jack' 45 'Florida']]


In [8]:
## Create a Data frame From a List of Dictionaries

data=[
    {'Name':'Krish','Age':32,'City':'Bangalore'},
    {'Name':'John','Age':34,'City':'Bangalore'},
    {'Name':'Bappy','Age':32,'City':'Bangalore'},
    {'Name':'JAck','Age':32,'City':'Bangalore'}
    
]
df=pd.DataFrame(data)
print(df)
print(type(df))

    Name  Age       City
0  Krish   32  Bangalore
1   John   34  Bangalore
2  Bappy   32  Bangalore
3   JAck   32  Bangalore
<class 'pandas.core.frame.DataFrame'>


In [ ]:
df=pd.read_csv('sales_data.csv')
df.head(6)# first 6 rows exclyding the header row ofcourse

Transaction ID        Date Product Category                 Product Name  \
0           10001  2024-01-01      Electronics                iPhone 14 Pro   
1           10002  2024-01-02  Home Appliances             Dyson V11 Vacuum   
2           10003  2024-01-03         Clothing             Levi's 501 Jeans   
3           10004  2024-01-04            Books            The Da Vinci Code   
4           10005  2024-01-05  Beauty Products      Neutrogena Skincare Set   
5           10006  2024-01-06           Sports  Wilson Evolution Basketball   

   Units Sold  Unit Price  Total Revenue         Region Payment Method  
0           2      999.99        1999.98  North America    Credit Card  
1           1      499.99         499.99         Europe         PayPal  
2           3       69.99         209.97           Asia     Debit Card  
3           4       15.99          63.96  North America    Credit Card  
4           1       89.99          89.99         Europe         PayPal  
5           5       29.99         149.95           Asia    Credit Card

In [11]:
df.tail(5)

Transaction ID        Date Product Category  \
235           10236  2024-08-23  Home Appliances   
236           10237  2024-08-24         Clothing   
237           10238  2024-08-25            Books   
238           10239  2024-08-26  Beauty Products   
239           10240  2024-08-27           Sports   

                                        Product Name  Units Sold  Unit Price  \
235  Nespresso Vertuo Next Coffee and Espresso Maker           1      159.99   
236                        Nike Air Force 1 Sneakers           3       90.00   
237           The Handmaid's Tale by Margaret Atwood           3       10.99   
238             Sunday Riley Luna Sleeping Night Oil           1       55.00   
239                       Yeti Rambler 20 oz Tumbler           2       29.99   

     Total Revenue         Region Payment Method  
235         159.99         Europe         PayPal  
236         270.00           Asia     Debit Card  
237          32.97  North America    Credit Card  
238          55.00         Europe         PayPal  
239          59.98           Asia    Credit Card

In [10]:
data=[
    {'Name':'Krish','Age':32,'City':'Bangalore'},
    {'Name':'John','Age':34,'City':'Bangalore'},
    {'Name':'Bappy','Age':32,'City':'Bangalore'},
    {'Name':'JAck','Age':32,'City':'Bangalore'}
    
]
df =pd.DataFrame(data)

In [11]:
### Accessing Data From Dataframe
df

Name  Age       City
0  Krish   32  Bangalore
1   John   34  Bangalore
2  Bappy   32  Bangalore
3   JAck   32  Bangalore

In [12]:
print(df['Name'])#This now returns a series
print(type(df['Name']))


0    Krish
1     John
2    Bappy
3     JAck
Name: Name, dtype: object
<class 'pandas.core.series.Series'>


In [11]:
df.loc[0]["Name"]

'Krish'

In [13]:
df.loc[0,"Name"]

'Krish'

In [15]:
df.iloc[0]

Name        Krish
Age            25
City    Bangalore
Name: 0, dtype: object

In [14]:
df.iloc[0,0]

'Krish'

In Pandas, `.loc[]` and `.iloc[]` are used to access rows and columns of a DataFrame, but they differ in how they interpret the indexing:

---

### **1. `df.loc[]`**: Label-Based Indexing
- **Usage**: Selects data by the labels of rows and columns.
- **Indexes**: Works with row and column labels, not numerical positions.
- **Inclusive**: Both start and end labels are included in slicing.
  
#### Examples:
```python
import pandas as pd

# Sample DataFrame
data = {'A': [10, 20, 30], 'B': [40, 50, 60]}
df = pd.DataFrame(data, index=['row1', 'row2', 'row3'])

# Access a single row by label
print(df.loc['row1'])  
# Output: A    10
#         B    40

# Access multiple rows by labels
print(df.loc[['row1', 'row3']])  
# Output:
#          A   B
# row1    10  40
# row3    30  60

# Access with a slice of labels
print(df.loc['row1':'row3'])  
# Output:
#          A   B
# row1    10  40
# row2    20  50
# row3    30  60

# Access specific row and column by label
print(df.loc['row1', 'A'])  # Output: 10
```

---

### **2. `df.iloc[]`**: Integer-Based Indexing
- **Usage**: Selects data by the numerical positions of rows and columns.
- **Indexes**: Works with 0-based numerical positions.
- **Exclusive**: End position is excluded in slicing.

#### Examples:
```python
# Access a single row by index position
print(df.iloc[0])  
# Output: A    10
#         B    40

# Access multiple rows by index positions
print(df.iloc[[0, 2]])  
# Output:
#          A   B
# row1    10  40
# row3    30  60

# Access with a slice of positions
print(df.iloc[0:2])  
# Output:
#          A   B
# row1    10  40
# row2    20  50

# Access specific row and column by position
print(df.iloc[0, 0])  # Output: 10
```

---

### Key Differences:

| Feature         | `df.loc[]`                       | `df.iloc[]`                     |
|------------------|----------------------------------|----------------------------------|
| **Indexing Type** | Label-based indexing            | Integer-based indexing          |
| **Start/End**    | Inclusive of end label           | Exclusive of end position       |
| **Index Type**   | Uses row/column names            | Uses 0-based integer positions  |
| **Error Handling** | KeyError for invalid labels    | IndexError for invalid positions|

#### **Practical Usage:**
- Use **`loc`** when you know the labels (e.g., `'row1'`, `'A'`).
- Use **`iloc`** when you know the positions (e.g., `0`, `1`).



In [34]:
df

Name  Age       City
0  Krish   25  Bangalore
1   John   30   New York
2   Jack   45    Florida

In [35]:
## Accessing a specified element
df.at[2,'Age']

45

In [ ]:
#FOR LOC or ILOC the column was not required arguemnt but for AT it is required
df.at[2]['Age']#for at column is a required argument

TypeError: DataFrame._get_value() missing 1 required positional argument: 'col'

In [36]:
df.at[2,'Name']

'Jack'

In [38]:
## Accessing a specified element using iat
df.iat[2,2]

'Florida'

In [39]:
df

Name  Age       City
0  Krish   25  Bangalore
1   John   30   New York
2   Jack   45    Florida

In [40]:
### Data MAnipulation with Dataframe
df

Name  Age       City
0  Krish   25  Bangalore
1   John   30   New York
2   Jack   45    Florida

In [41]:
## Adding a column
df['Salary']=[50000,60000,70000]#Adds this column to the original dataframe
df

Name  Age       City  Salary
0  Krish   25  Bangalore   50000
1   John   30   New York   60000
2   Jack   45    Florida   70000

In [46]:
## Remove a column
#axis=0 checks for "Salary" in the row index
#axis=1 checks for "Salary" in the column index
#without inplace=True, the original dataframe is not modified ,it just returns a new dataframe with the column removed
df.drop('Salary',axis=1,inplace=True)


In [47]:
df

Name  Age       City
0  Krish   25  Bangalore
1   John   30   New York
2   Jack   45    Florida

In [48]:
## Add age to the column
df['Age']=df['Age']+1
df

Name  Age       City
0  Krish   26  Bangalore
1   John   31   New York
2   Jack   46    Florida

In [51]:
df.drop(0,inplace=True)

In [52]:
df

Name  Age      City
1  John   31  New York
2  Jack   46   Florida

In [53]:
df=pd.read_csv('sales_data.csv')
df.head(5)

Transaction ID        Date Product Category             Product Name  \
0           10001  2024-01-01      Electronics            iPhone 14 Pro   
1           10002  2024-01-02  Home Appliances         Dyson V11 Vacuum   
2           10003  2024-01-03         Clothing         Levi's 501 Jeans   
3           10004  2024-01-04            Books        The Da Vinci Code   
4           10005  2024-01-05  Beauty Products  Neutrogena Skincare Set   

   Units Sold  Unit Price  Total Revenue         Region Payment Method  
0           2      999.99        1999.98  North America    Credit Card  
1           1      499.99         499.99         Europe         PayPal  
2           3       69.99         209.97           Asia     Debit Card  
3           4       15.99          63.96  North America    Credit Card  
4           1       89.99          89.99         Europe         PayPal

In [54]:
# Display the data types of each column
print("Data types:\n", df.dtypes)#property

# Describe the DataFrame
print("Statistical summary:\n", df.describe())#function that returns a dataframe --->only works for numerical columns



Data types:
 Transaction ID        int64
Date                 object
Product Category     object
Product Name         object
Units Sold            int64
Unit Price          float64
Total Revenue       float64
Region               object
Payment Method       object
dtype: object
Statistical summary:
        Transaction ID  Units Sold   Unit Price  Total Revenue
count       240.00000  240.000000   240.000000     240.000000
mean      10120.50000    2.158333   236.395583     335.699375
std          69.42622    1.322454   429.446695     485.804469
min       10001.00000    1.000000     6.500000       6.500000
25%       10060.75000    1.000000    29.500000      62.965000
50%       10120.50000    2.000000    89.990000     179.970000
75%       10180.25000    3.000000   249.990000     399.225000
max       10240.00000   10.000000  3899.990000    3899.990000


The values 25%, 50%, and 75% in the summary statistics table represent **percentiles** for the given data column. These values divide the data into quarters, providing insight into its distribution:

1. **25% (First Quartile, Q1):**
   - This is the value below which 25% of the data lies. 
   - It gives the lower bound of the middle 50% of the data.
   - Example: For the `Units Sold` column, the 25% value is **1.000000**, meaning 25% of transactions had 1 unit sold or fewer.

2. **50% (Median, Q2):**
   - This is the middle value when the data is sorted, also known as the **median**.
   - Half the data lies below this value.
   - Example: For the `Unit Price` column, the median is **89.990000**, meaning half of the unit prices are below or equal to this value.

3. **75% (Third Quartile, Q3):**
   - This is the value below which 75% of the data lies.
   - It gives the upper bound of the middle 50% of the data.
   - Example: For the `Total Revenue` column, the 75% value is **399.225000**, meaning 75% of the transactions have total revenue below this value.

### Importance
Percentiles are useful for understanding the distribution of data:
- **25%**: Helps identify the lower range of the data.
- **50%**: Provides the central tendency (median).
- **75%**: Shows the upper range of most data points. 

Together, these percentiles help identify outliers and skewness in the data.

In [55]:
df.describe()

Transaction ID  Units Sold   Unit Price  Total Revenue
count       240.00000  240.000000   240.000000     240.000000
mean      10120.50000    2.158333   236.395583     335.699375
std          69.42622    1.322454   429.446695     485.804469
min       10001.00000    1.000000     6.500000       6.500000
25%       10060.75000    1.000000    29.500000      62.965000
50%       10120.50000    2.000000    89.990000     179.970000
75%       10180.25000    3.000000   249.990000     399.225000
max       10240.00000   10.000000  3899.990000    3899.990000